### DISCUSSION TITLE
Performance implication of skipped tasks? Interaction with TaskGroup?

### DISCUSSION THREAD

**Discussion by p-hebert** [apache/airflow#21739] [Q&A] [answered]
Hi there!

We are currently refactoring one of our DAGs to enhance its functionalities.
One of the consequences of such refactoring is that this once generalized DAG will now have more specific branches for each of the use cases (different reports generated).
Our current strategy is to branch into several TaskGroups and to execute only the TaskGroup relevant for the report type specified in the DagRun conf.

The issue we are worried about is overburdening the scheduler as each of the branches will be comprised of several tasks (in the 10s). Reports are generated in batch (10-15 DagRuns at a time).
We have a small team (2 of us to manage a full E2E ETL + application), and thus we are ready to go a bit more on the over-optimization side than the sub-optimization side to avoid potentially time-consuming failures.
We are running on a K8S cluster with HPA enabled.

Thus my questions:

1. How heavy is a skipped task on the scheduler?
2. Is this even a concern we should have?
3. Does TaskGroups provide an optimization advantage for the scheduler if they are skipped (as opposed to skipping tasks individually?)
4. Would we be better to split off each report type in their own DAG and call instead of having several TaskGroups in a single DAG?

**Comment by potiuk** [comment, marked as answer]
I think it's better to split. It should have no "real" influence if you have 10-15 dags (Airflow can handle 1000s) but it's generally better to use the "pythonic" approach of DAG creation and use common "shared" code to generate multiple dags rather than rely on the branch operators for the logic. 

Generally - follow https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html to write your DAGs and you will be fine.

**Reply by p-hebert** [reply]
Great; I'll opt for that solution then. Thanks again for your answer & good work Jarek!
